rest/interceptors/src/main/java/com/connexta/spring/interceptor/VersionInterceptor.java

```diff
@@ -0,0 +1,290 @@
+/**
+ * Copyright (c) Connexta
+ *
+ * <p>This is free software: you can redistribute it and/or modify it under the terms of the GNU
+ * Lesser General Public License as published by the Free Software Foundation, either version 3 of
+ * the License, or any later version.
+ *
+ * <p>This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details. A copy of the GNU Lesser General Public
+ * License is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package com.connexta.spring.interceptor;
+
+import com.connexta.discovery.rest.models.ErrorMessage;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import java.io.IOException;
+import java.io.InputStream;
+import java.lang.reflect.Method;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.security.CodeSource;
+import java.security.ProtectionDomain;
+import java.util.Map;
+import java.util.Properties;
+import java.util.concurrent.ConcurrentHashMap;
+import javax.annotation.Nullable;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import org.apache.commons.io.IOUtils;
+import org.apache.commons.io.input.ProxyInputStream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.http.MediaType;
+import org.springframework.stereotype.Component;
+import org.springframework.web.method.HandlerMethod;
+import org.springframework.web.servlet.HandlerInterceptor;
+import org.springframework.web.servlet.ModelAndView;
+
+/**
+ * Spring interceptor capable of validating the version provided by the client with the version
+ * implemented by the server and return an error message accordingly. It will also automatically add
+ * the <code>Content-Version</code> header with the server version for the API called. This only
+ * happens in cases where the controller was called and generated a response and no other
+ * interceptors failed.
+ *
+ * <p>The server version for the API is retrieved from the dependency.proeprties file that must be
+ * included in the generated spring API using Apache's ServiceMix depends maven plugin.
+ */
+@Component
+public class VersionInterceptor implements HandlerInterceptor {
```

Comment:
❓Could we keep this simple for now and let the server handle this in its own repo until we know more about how we want this to work?
Or could we inject the API version in the handler doing something like this instead: https://docs.spring.io/spring-boot/docs/current/reference/html/howto-properties-and-configuration.html?

Reply:
I honestly prefer extracting the version dynamically from the API code as opposed to relying on the server providing it via injection. This way, we are guaranteed that we are comparing and advertising the right version of the spec. Less chance of forgetting to bump that version.

My preference would be for the generators to actually annotate the controller's interface with the version as it would make things much easier to get to it. For the time being, this works like a charm.

Reply:
The other point is this logic accounts for any combined APIs that are handled by a given spring boot application. In such case, you would have to handle multiple versions and keep a mapping of some sort as part of configuration. This relies on the actual handler that will be called to retrieve the version associated with the spec it was generated and overloaded from.